### PR TITLE
Bug 1600759 - Perfherder's Graph Alt. View for Screen Readers

### PR DESCRIPTION
**Pull request by yogmel**
## Description of issue
Perfherder's Graph View contains graphs and elements which are not screen reader accessible.

**Tasks**:
- [X] Get Graph data into table;
- [X] Highlight Revision;
- [x] Add Tooltip information;
- [ ] When highlight, show and focus highlighted row;
- [ ] Toogle column's visibility;
- [x] Add tests.

Next PR:
-  (Needs investigation) Show trends and outliers;

## Proposed Solution
Perfherder's Graph View, with the Table showing. On the left, there are legend cards, with all tests information. Turned on tests have distinct colors, such as blue and red and different icons. On the right, there is first a row of buttons: "Last day", "Add test data" and "Graphs View". Below that, there is the table, with navigation on top. The first column refers to the revision test date. The other columns correspond to each test data. There 7 columns in total (One is the date column and the other six are tests). Some rows have a yellow background. Below the table, there are two input fields and a "Highlight alerts" button.

- A **button** next to "Add test data" was added, to toggle between Graph and Table view.
- If there is any **highlighted revision**, its table row will have a yellow background and an `aria-label` of "highlighted revision".

- Users can **show and hide test** via Checkbox in Legend Card:
The same table as the first screenshot, but the first two Legend cards are turned off, that is, they are grey. This reflects in the table, which now has 5 columns instead of 7.

## Testing
- Compared some data points in the graph to the ones shown in table cells.
- Turned highlighted revisions on and off and check if it propagated both in the graph and table.
- Still was not able to test with SR. I will edit here once it's done.

### CODE REVIEW

**Comment by yogmel** [comment]
Hi @camd and @sarah-clements ,
I have some comments:
- I separated in commits to facilitate review;
- `GraphsContainer` make it look like a lot of changes were done, but I added only a `{!showTable && ()}` around the component. The other changes was made by `yarn lint` I suppose.
- I couldn't figure out how to get links and content from `GraphTooltip`, so I will need some help in this.

Thanks!

**Comment by yogmel** [comment]
Hi @camd, I still didn't talk to him about this PR, but I would love to do it.
I will try to schedule a call with him and ask someone to deploy this at prototype, if possible.

**Comment by MarcoZehe** [comment]
Note that from a markup point of view, this looks good to me.

**Comment by sarah-clements** [comment]
Great start! A few comments on the layout/visuals:

1) I don't think we need the legend cards on the left, since any relevant information is already in the column headers. Whatever is missing can be added (such as the framework, but I dont't think we need the signature hash).

2) We don't need to make this as visually colorful as the graph. To keep it simple - and since the primary target is screen reader users - I think we should omit the colors to differentiate the tests.

3) For the highlighting of revisions, please make it a color that matches with the secondary or info colors (like a light info color that's used for Alerts) and have it go all the way across the column.

File: ui/css/perf.css [hunk 1]
```diff
@@ -491,3 +491,42 @@ li.pagination-active.active > button {
...
   visibility: visible;
 }
+
+/* table in graphs view */
+.table-header {
+  white-space: pre-line !important;
+  word-wrap: break-word;
+  display: flex;
+  justify-content: center;
+  align-items: center;
```
Comment (sarah-clements) [inline ui/css/perf.css]: Could you use bootstrap utility classes for this instead?

File: ui/perfherder/graphs/GraphsViewControls.jsx [hunk 2]
```diff
@@ -80,6 +82,13 @@ export default class GraphsViewControls extends React.Component {
               Add test data
             </Button>
           </Col>
+          {!hasNoData && (
+            <Col sm="auto" className="p-2">
+              <Button color="darker-info" onClick={toggleTableView}>
+                {showTable ? 'Graphs View' : 'Table View'}
+              </Button>
+            </Col>
+          )}
```
Comment (sarah-clements) [inline ui/perfherder/graphs/GraphsViewControls.jsx]: You can move this to line 101, including it in the second condition of the ternary statement that begins on line 94 (if `hasNoData` is true, do this, otherwise do that (starting on line 101).

File: ui/perfherder/graphs/TableView.jsx [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        ),
+        Cell: ({ original }) => {
+          return <div>{original[value]}</div>;
+        },
+        headerClassName: `table-header ${item.color[0]}`,
+        getProps: (state, rowInfo) => ({
+          style: {
+            background: rowInfo && rowInfo.original.highlighted && '#ffde2f',
+          },
+          'aria-label': 'highlighted revision',
```
Comment (sarah-clements) [inline ui/perfherder/graphs/TableView.jsx]: A thought I had... if someone enters a revision, should the table results filter to that revision(s) or otherwise jump to it first - since that is what someone would want to know about?

**Comment by ionutgoldan** [comment]
@bebef1987 @marianrai @alexandru-io please do a local check for this PR, so I know you're ok with the changes done in Graphs view, on the `legend` component.

**Comment by bebef1987** [comment]
Please add Commit information and a link to treeheder. 
The current table format is not usable as the graph view

**Comment by alexandru-io** [comment]
1. I would place the Table/Graph view toggle before the Time range button and eventually search for a more representative toggle button graphic.
Screenshot from 2020-02-17 11-20-09
2. I would replace the **Date** column with a **Commit** one, with a link to pushlog and jobs, just in the graph view. Eventually include in a tool-tip or parenthesis the date of the revision.
Screenshot from 2020-02-17 11-29-36
3. I would use another (eye-comforting) color to highlight the revisions and include the first column also in the highlight. I would also improve the number formatting for a better readability and the measurement unity (3151094.00 -> 3,151,094.00)
Screenshot from 2020-02-17 11-31-19
4. I would use the empty row in the middle of the data columns on the top to represent to color of the test data and format the signature as in the legend card:
```
|   color   |
| Test name |
| repo      |
| platform  |
```
Screenshot from 2020-02-17 11-37-44
Screenshot from 2020-02-17 11-37-33

5. The color difference between Previous/Next button when they're available/not is too small as they they're both available they look much like they are not.
Previous = unavailable, Next = available
Screenshot from 2020-02-17 11-44-04
Previous = available, Next = available
Screenshot from 2020-02-17 11-44-12

This looks like a great work and I didn't find functionality issues yet, but in terms of usability the fixes below seem fair to me.

File: ui/perfherder/graphs/GraphsViewControls.jsx (hunk 2 above)
Comment (yogmel) [reply ui/perfherder/graphs/GraphsViewControls.jsx]: Hmm I have changed the button position per @alexandru-io's request and applying this suggestion now would alter its ordering.

File: ui/perfherder/graphs/TableView.jsx (hunk 3 above)
Comment (yogmel) [reply ui/perfherder/graphs/TableView.jsx]: That's a good idea. I think the second suggestion would be the best one. I took note of that and I will try to implement it.

File: ui/perfherder/graphs/GraphsViewControls.jsx (hunk 2 above)
Comment (sarah-clements) [reply ui/perfherder/graphs/GraphsViewControls.jsx]: If it's no longer applicable, that's fine :)

**Comment by yogmel** [comment]
This last commit adds most of the suggestions made above.

The information from the tooltip was inserted. The "job, compare" links only appear when there is a mouse hover or a keyboard focus.

There are still some additional tasks and tests to be made that I will update on the PR description.

Screenshot of Graphs's Table View. From top to bottom, there is a row of buttons: "Graphs View", "Last 14 days", "Add test data". Down below, there is a table with navigation buttons. There are 4 columns and 10 rows (except for table header). The first table header is Date. The following ones are related to tests - there is a color border on top, the test name, a hint (lower is better), and the platform. The rows contain the number, with the delta on the side. Down below the row content, there is the job and compare links. Below the table, there are 2 inputs and a button "Highlight Alerts"

**Comment by alexandru-io** [comment]
@yogmel this looks much better.
I tested around some scenarios and there are few issues:
- when i tried to add a new test data in table view I am getting the error: `Something went wrong TypeError: original[deltaValueKey] is undefined`, but not for all tests. Couldn't identify for what tests. For example I tried to add a talos test on a graph with raptor. But only table view throws this error. After `F5` the page is loaded successfully on graph view, but it's thrown back when toggling to table view.
- Not sure how the data is sorted, but something doesn't seem right. In the screesnhot below the dates end at Feb 20 (today) and continues from Jan 21 next hour after the previous from Feb 20. According to the timerange filter, the data should indeed stop on Jan 21, but the ascending/descending date sorting should be (Jan 21 ... Feb 20)/(Feb 20 ... Jan 21). Guess you need to change the data type of the filter.
Screenshot from 2020-02-20 15-50-13

Not sure whether you got to work on this or you're going to, but the information about repo and platform still looks "unfinished".

Overall this is a good piece of work. Few more fixes and we're good to go.

**Comment by yogmel** [comment]
In this latest push:

- Add some **tests**: `table_view_text.jsx` and I had to create a mock data `performance_testData.json`.

- Jamie's request (**Screen-Reader improvements**):
  - Job and compare links are not SR accessible while hidden. For that reason, I made them appear all the time, but with low opacity. When focused or hovered, the links have full opacity (file: `perf.css`);
  - "Table View" button is not the first element on DOM. The proposed solution here was to insert a `sr-only` button before the Legend Cards (file: `GraphsView.jsx`).

- Add test Badge, for example "talos", "raptor" (file: `TableView.jsx`).

Screenshot of Graphs's Table View. From top to bottom, there is a row of buttons: Graphs View, Last 14 days, Add test data. Down below, there is a table with navigation buttons. There are 4 columns and 10 rows (except for table header). The first table header is Date. The following ones are related to tests - there is a color border on top, the test name, a hint (lower is better), a test badge, and the platform and repo. The rows contain the number, with the delta on the side. Down below the row content, there is the job and compare links.

In `TableView.jsx`:
- Refactor:
  - Separate into different functions;
  - Optimized data structure;

- Fix date sorting
- Fix "add test data" error: @alexandru-io I think that error was triggered when tests with empty cells were chosen. If that's not the case, and the problem still persists, please let me know.

**Comment by yogmel** [comment]
> Not sure whether you got to work on this or you're going to, but the information about repo and platform still looks "unfinished".

@alexandru-io Thanks for your input!
Do you mean the style or the functionality? I realized later that users can click on repo and platform names on the Legend Card and it will open the "add test modal". Is that what you meant?

**Comment by alexandru-io** [comment]
> > Not sure whether you got to work on this or you're going to, but the information about repo and platform still looks "unfinished".
> 
> @alexandru-io Thanks for your input!
> Do you mean the style or the functionality? I realized later that users can click on repo and platform names on the Legend Card and it will open the "add test modal". Is that what you meant?

I'm talking about the styling. I don't think is bringing any value/it's not relevant popping the add test modal when clicking on that.

**Comment by sarah-clements** [comment]
> > > Not sure whether you got to work on this or you're going to, but the information about
 
> I'm talking about the styling. I don't think is bringing any value/it's not relevant popping the add test modal when clicking on that.

I think the styling looks fine. The purpose of it is primarily for screen reader users so I do not think it needs to be as visually dynamic as the main graphs view.

File: tests/ui/mock/performance_testData.json [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    "visible": true,
+    "name": "tart_flex opt e10s stylo",
+    "signature_id": 1647462,
+    "signatureHash": "664c274bfcf996b14aa7771b6f2779531c309868",
+    "framework_id": 1,
+    "platform": "linux64",
+    "repository_name": "mozilla-central",
+    "projectId": 1,
+    "id": "mozilla-central tart_flex opt e10s stylo",
+    "data": [
```
Comment (sarah-clements) [inline tests/ui/mock/performance_testData.json]: Instead of adding all of this data, I think you can use the `createGraphData` function with existing test data. Look in the tests/ui/perfherder/graphs_view_test file.

File: ui/perfherder/graphs/GraphsView.jsx [hunk 5]
```diff
@@ -346,7 +346,18 @@ class GraphsView extends React.Component {
               <Col
                 className={`${testData.length ? 'graph-chooser' : 'col-12'}`}
               >
-                <Container className="graph-legend pl-0 pb-4">
+                <button
```
Comment (sarah-clements) [inline ui/perfherder/graphs/GraphsView.jsx]: Use the reactstrap Button component and then you don't need to add the type.

Comment (sarah-clements) [reply ui/perfherder/graphs/GraphsView.jsx]: Although, I suppose it doesn't matter here since it's for screen readers... :)

File: ui/perfherder/graphs/TableView.jsx [hunk 6]
```diff
@@ -1,21 +1,77 @@
...
+        ? item.data[prevFlotDataPointIndex].y
+        : valueY;
+    const deltaValue = valueY - v0;
+    const deltaPercent = valueY / v0 - 1;
+
+    // get push link
+    let prevRevision;
+    let pushUrl;
+    if (prevFlotDataPointIndex !== -1) {
+      prevRevision = item.data[prevFlotDataPointIndex].revision;
```
Comment (sarah-clements) [inline ui/perfherder/graphs/TableView.jsx]: Some of this logic from the tooltip can be put into a helper function to be used by both components. That way if it needs to be changed in the future, you only have to change it in that helper function.

**Comment by sarah-clements** [comment]
This looks good but let's focus on getting this approved and landed. You can do follow up improvements and features in a separate pull request, otherwise this will get too big.

File: ui/perfherder/graphs/TableView.jsx [hunk 7]
```diff
@@ -28,169 +84,177 @@ const TableView = ({
...
-              item => item && dataPoint.revision.includes(item),
+        // populate array with only data points
+        item.data.forEach((dataPoint, dataIndex) => {
+          allDataPoints.push(
+            getRowDataPoint(
+              item,
+              dataIndex,
+              dataPoint,
+              highlightAlerts,
+              highlightedRevisions,
```
Comment (sarah-clements) [inline ui/perfherder/graphs/TableView.jsx]: What changed in this file from the last commit?

File: ui/perfherder/graphs/TableView.jsx (hunk 6 above)
Comment (alexandru-io) [reply ui/perfherder/graphs/TableView.jsx]: @yogmel, @sarah-clements  just be aware that there's an [open PR 6066](https://github.com/mozilla/treeherder/pull/6066/files) that is fixing the [empty pushlogs](https://hg.mozilla.org/integration/autoland/pushloghtml?fromchange=458b5c5e018362ad113be1baf2feb4bdc2913666&tochange=458b5c5e018362ad113be1baf2feb4bdc2913666) from [graph tooltip](https://treeherder.mozilla.org/perf.html#/graphs?highlightAlerts=1&selected=1926849,1057676531&series=autoland,1926849,1,1&timerange=1209600&zoom=1582199603282,1582307149191,312.513848695788,364.11166716660455) that appears from the second data-point (first retrigger) of a revision. This is because `prevRevision` is taking the `revision` hash from `prevFlotDataPointIndex`, and if the revision is the same, the pushlog link generated is not correct.

Moving this to a helper function might make fixing this bug harder if it doesn't pass `dataPointDetails` and `testDetails` as params. I don't mind either if you include the fix into this PR. Whatever works best for you. Thanks

File: ui/perfherder/graphs/TableView.jsx (hunk 7 above)
Comment (yogmel) [reply ui/perfherder/graphs/TableView.jsx]: Some functions were created and I changed the data structured sent to the table. I basically removed [this](https://github.com/mozilla/treeherder/pull/6001/commits/6183d6c3115cb090d203639a60101a071715ad4d#diff-fce1c2201d8b61dea9a7b9418006ace2R60-R63) and moved it into an object. I thought it was cleaner.

[Before](https://github.com/mozilla/treeherder/pull/6001/commits/6183d6c3115cb090d203639a60101a071715ad4d#diff-fce1c2201d8b61dea9a7b9418006ace2R160-R174) | [Current](https://github.com/mozilla/treeherder/pull/6001/files#diff-fce1c2201d8b61dea9a7b9418006ace2R161-R177)

(there are some lines highlighted, but apparently Github doesn't jump to them when you click the link)

File: ui/perfherder/graphs/GraphsView.jsx (hunk 5 above)
Comment (yogmel) [reply ui/perfherder/graphs/GraphsView.jsx]: I'll change it, as we are getting another update on this :)

Comment (yogmel) [reply ui/perfherder/graphs/GraphsView.jsx]: And sorry I always forget to add as a Reactstrap button.

**Comment by yogmel** [comment]
I have implemented @sarah-clements's suggestions.

About the separation of helpers' functions, I can wait for @alexandru-io merge or implement in the follow-up PR.

**Comment by sarah-clements** [comment]
I tested this andI noticed that the 'highlight alerts' button next to the input boxes is controlling the highlighting for the revisions. Those are two separate features (check the implementation in the graphs view). So the toggle for highlight alerts should *not* control whether the revisions are highlighted. If you want to see a highlighted alert, go to the alerts view and hover over a test - click on the graphs link (the effect on the datapoint is the same).  If we don't have the highlight alert functionality implemented in this pr, hide the button when in table view for now.

Also, please file new bugs for the additional features to be added. @yogmel you can assign yourself to them, but we should have bugs filed.

**Comment by yogmel** [comment]
@sarah-clements I have found the piece of code that gets the query params, but I would like to discuss options before making a move. This latest push hides the highlight alerts input in the table view.
I'll file all bugs at the start of next week. :)

File: ui/perfherder/graphs/GraphsViewControls.jsx [hunk 8]
```diff
@@ -113,38 +113,40 @@ export default class GraphsViewControls extends React.Component {
...
-                  onClick={() =>
-                    updateStateParams({ highlightAlerts: !highlightAlerts })
-                  }
-                  active={highlightAlerts}
-                >
-                  Highlight alerts
-                </Button>
-              </Col>
-            </Row>
+            {!showTable && (
```
Comment (sarah-clements) [inline ui/perfherder/graphs/GraphsViewControls.jsx]: So we just want to hide the "highlight alerts" button in the table view until that logic is fixed.

**Comment by yogmel** [comment]
@sarah-clements ok, changed to hide only the "Highlight Alerts" button. Input fields are still shown and highlight cells.

**Comment by alexandru-io** [comment]
Yep, looks good. Even the `platform | repo` thing looks better now with that nuance of gray. Good job, @yogmel !

**Comment by yogmel** [comment]
Thank you, everyone :)
